### PR TITLE
fix(metrics): Updated how we enable/check metrics in React app

### PR DESCRIPTION
**Pull request by vbudhram**
## Because

- There are certain scenarios where might ignore metrics prefs

## This pull request

- Standard how we perform this check

## Issue that this pull request solves

Closes: https://mozilla-hub.atlassian.net/browse/FXA-9403

## Checklist

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Other information (Optional)

Targetting this as a point release

### CODE REVIEW

File: packages/fxa-settings/src/components/App/index.tsx [hunk 1]
```diff
@@ -92,16 +92,27 @@ export const App = ({
...
-    // Don't initialize until integration is ready!
+  const getMetricsEnabled = useCallback(() => {
+    if (metricsLoading || !integration || isSignedIn === undefined) {
+      return;
+    }
+    return data?.account?.metricsEnabled || !isSignedIn;
+  }, [metricsLoading, integration, isSignedIn, data?.account?.metricsEnabled]);
+  const metricsEnabled = getMetricsEnabled();
+
+  useEffect(() => {
```
Comment (chenba) [inline packages/fxa-settings/src/components/App/index.tsx:103]: We don't want to use useEffect as it executes at the end of rendering so this won't execute until all of the children finish rendering but we might want Glean to emit events when the children render.

Comment (vbudhram) [reply packages/fxa-settings/src/components/App/index.tsx:103]: @chenba I noticed using the `useMemo` it initalizes Glean several times, at least with console logging.

Screenshot 2024-04-11 at 2 12 08 PM

Are there certain events that need to be initalized before rendering?

Comment (chenba) [reply packages/fxa-settings/src/components/App/index.tsx]: > useMemo it initalizes Glean several times

Maybe dependencies for useMemo changed?  I don't think it's harmful to init glean.js more than once though.

> Are there certain events that need to be initalized before rendering?

Most if not all of the `.view` events are called in a `useEffect` of their component.

File: packages/fxa-settings/src/components/App/index.tsx [hunk 2]
```diff
@@ -92,16 +99,23 @@ export const App = ({
 
   const isSignedIn = isSignedInData?.isSignedIn;
 
+  const getMetricsEnabled = useCallback(() => {
+    if (metricsLoading || !integration || isSignedIn === undefined) {
+      return;
+    }
+    return data?.account?.metricsEnabled || !isSignedIn;
```
Comment (chenba) [inline packages/fxa-settings/src/components/App/index.tsx:106]: This logic looks identical to the [old line 104](https://github.com/mozilla/fxa/pull/16696/files#diff-36aa53053f1cbb57d4a94e4cf502a82d89c9ef2cfdb037ffc6e31ad298c74893L104) below, so I don't think I understand how this changes the handling of state change of `isSignedIn` while we still don't have a verified session to get the actual user opt-out pref.

**Review by chenba — APPROVED** [review]
wfm

**Review by LZoog — APPROVED** [review]
I tested this locally on `main` and saw `glean.setEnabled` called with `true` in React when reaching signup, and then called with `false` after signup when navigating to `confirm_signup_code`.

In your branch, it doesn't get recalled with `false` on confirm signup code. 👍

**Pull request merged**